lib/index.ts

```diff
@@ -15,6 +15,37 @@ const cancelableClient: CancelableAxiosInstance = Object.assign(client, {
 	CancelToken: Axios.CancelToken,
 	isCancel: Axios.isCancel,
 })
+cancelableClient.interceptors.response.use(r => r, async (error) => {
+	const { config, response, request: { responseURL } } = error
+	const { status, headers } = response
+	const originalRequest = config
+
+	/**
+	 * Retry requests if they failed due to maintenance mode
+	 *
+	 * The delay is exponential. It starts at 2s and then doubles
+	 * until a final retry after 32s. This results in roughly 1m of
+	 * retries until we give up and throw the axios error towards
+	 * the caller.
+	 */
+	if (status === 503
+		&& headers['x-nextcloud-maintenance-mode'] === '1'
+		&& config.retryIfMaintenanceMode
+		&& (!config._retryDelay || config._retryDelay <= 32)) {
+		const retryDelay = (config._retryDelay ?? 1) * 2
+		console.warn(`Request to ${responseURL} failed because of maintenance mode. Retrying in ${retryDelay}s`)
+		await new Promise((resolve, _) => {
+			setTimeout(resolve, retryDelay*1000)
+		})
+
+		return cancelableClient({
+			...originalRequest,
+			_retryDelay: retryDelay,
```

Comment:
We can use [symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) to prevent external access of this value.


```ts
const RETRY_DELAY_KEY = Symbol('retryDelay')

// ---

&& (!config[RETRY_DELAY_KEY] || config[RETRY_DELAY_KEY] <= 32)) {
const retryDelay = (config[RETRY_DELAY_KEY] ?? 1) * 2

// --

return cancelableClient({
  ...config,
  [RETRY_DELAY_KEY]: retryDelay,
})

```